File: server/src/pkm/server/fts.py
# pattern: Functional Core
"""Escape untrusted text into FTS5 MATCH expressions."""
from __future__ import annotations

import re


def _quote(term: str) -> str:
    return '"' + term.replace('"', '""') + '"'


def escape_fts_query(q: str) -> str:
    terms = [t for t in re.split(r"\s+", q.strip()) if t]
    if not terms:
        return '""'
    quoted = [_quote(t) for t in terms]
    quoted[-1] += "*"
    return " ".join(quoted)


def phrase_query(q: str) -> str:
    return _quote(q.strip())

File: web/src/replica/localApi/fts.ts
// pattern: Functional Core
// Port of server fts.py: escape untrusted text into FTS5 MATCH expressions.

const quote = (term: string): string => `"${term.replaceAll('"', '""')}"`;

export function escapeFtsQuery(q: string, exact = false): string {
  const terms = q.trim().split(/\s+/).filter((t) => t.length > 0);
  if (terms.length === 0) return '""';
  const quoted = terms.map(quote);
  if (!exact) quoted[quoted.length - 1] += "*";
  return quoted.join(" ");
}

export function phraseQuery(q: string): string {
  return quote(q.trim());
}
